File: python_archive/src/core/game_state.py
"""
GameState class with single state and efficient undo functionality.
Critical for high-performance MCTS implementation.
"""

from typing import List, Tuple, Optional
from .bitboard import BitBoard
from .move_delta import MoveDelta

class GameState:
    """
    Pente game state using single BitBoard with efficient make/undo operations.
    
    This is the core of the high-performance MCTS implementation:
    - Never copies state (eliminates thousands of deep copies per second)
    - Uses move deltas for perfect undo operations
    - Orders of magnitude faster than state copying approach
    """
    
    def __init__(self, board_size: int = 19, captures_to_win: int = 5, tournament_rule: bool = True):
        self.board_size = board_size
        self.captures_to_win = captures_to_win
        self.tournament_rule = tournament_rule
        
        # Single BitBoard instance - never copied!
        self.board = BitBoard()
        
        # Game state
        self.current_player = 1
        self.move_history: List[MoveDelta] = []
        self.captures = {1: 0, -1: 0}  # Number of pairs captured
        
        # Performance tracking
        self.move_count = 0
    
    def make_move(self, position: Tuple[int, int]) -> MoveDelta:
        """
        Make a move and return delta for undo.
        This is the critical path for MCTS performance.
        """
        row, col = position
        
        if not self.board.is_empty(row, col):
            raise ValueError(f"Position {position} is already occupied")
        
        # Place the stone
        self.board.set_stone(row, col, self.current_player)
        
        # Check for captures
        captured_positions = self._check_captures(row, col, self.current_player)
        captured_player = -self.current_player if captured_positions else 0
        capture_count = len(captured_positions) // 2  # Captures are in pairs
        
        # Update capture count
        if capture_count > 0:
            self.captures[self.current_player] += capture_count
        
        # Create move delta
        delta = MoveDelta(
            position=position,
            player=self.current_player,
            captured_positions=captured_positions,
            captured_player=captured_player,
            capture_count=capture_count
        )
        
        # Update state
        self.move_history.append(delta)
        self.move_count += 1
        self.current_player *= -1
        
        return delta
    
    def undo_move(self) -> Optional[MoveDelta]:
        """
        Undo the last move using the move delta.
        Returns the delta that was undone, or None if no moves to undo.
        """
        if not self.move_history:
            return None
        
        delta = self.move_history.pop()
        
        # Restore game state
        self.current_player = delta.player
        self.move_count -= 1
        
        # Remove the stone that was placed
        self.board.remove_stone(*delta.position)
        
        # Restore captured stones
        if delta.has_captures():
            for pos in delta.captured_positions:
                self.board.set_stone(*pos, delta.captured_player)
            self.captures[delta.player] -= delta.capture_count
        
        return delta
    
    def _check_captures(self, row: int, col: int, player: int) -> List[Tuple[int, int]]:
        """
        Check for captures after placing a stone.
        Returns list of captured stone positions.
        """
        directions = [(0, 1), (1, 0), (1, 1), (1, -1), (0, -1), (-1, 0), (-1, -1), (-1, 1)]
        captured_positions = []
        opponent = -player
        
        for dr, dc in directions:
            # Look for pattern: player -> opponent -> opponent -> player
            positions_to_check = []
            
            # Check first opponent stone
            r1, c1 = row + dr, col + dc
            if (0 <= r1 < self.board_size and 0 <= c1 < self.board_size and 
                self.board.get_stone(r1, c1) == opponent):
                positions_to_check.append((r1, c1))
                
                # Check second opponent stone
                r2, c2 = r1 + dr, c1 + dc
                if (0 <= r2 < self.board_size and 0 <= c2 < self.board_size and 
                    self.board.get_stone(r2, c2) == opponent):
                    positions_to_check.append((r2, c2))
                    
                    # Check for closing player stone
                    r3, c3 = r2 + dr, c2 + dc
                    if (0 <= r3 < self.board_size and 0 <= c3 < self.board_size and 
                        self.board.get_stone(r3, c3) == player):
                        # Capture found! Remove the two opponent stones
                        for cap_pos in positions_to_check:
                            self.board.remove_stone(*cap_pos)
                            captured_positions.append(cap_pos)
        
        return captured_positions
    
    def get_legal_moves(self) -> List[Tuple[int, int]]:
        """Get all legal moves for the current player."""
        moves = []
        
        # Tournament rule: Player -1's first move (second move of game) must be ≥3 from center
        if (self.tournament_rule and self.move_count == 1 and self.current_player == -1):
            center = self.board_size // 2
            for row in range(self.board_size):
                for col in range(self.board_size):
                    if self.board.is_empty(row, col):
                        distance = max(abs(row - center), abs(col - center))
                        if distance >= 3:
                            moves.append((row, col))
        else:
            # Normal case: all empty positions
            for row in range(self.board_size):
                for col in range(self.board_size):
                    if self.board.is_empty(row, col):
                        moves.append((row, col))
        
        return moves
    
    def is_terminal(self) -> bool:
        """Check if the game has ended."""
        return self.get_winner() is not None or len(self.get_legal_moves()) == 0
    
    def get_winner(self) -> Optional[int]:
        """
        Get the winner of the game.
        Returns 1, -1, or None (for ongoing game).
        """
        # Check capture win
        for player in [1, -1]:
            if self.captures[player] >= self.captures_to_win:
                return player
        
        # Check five-in-a-row
        return self._check_five_in_row()
    
    def _check_five_in_row(self) -> Optional[int]:
        """Check for five stones in a row."""
        directions = [(0, 1), (1, 0), (1, 1), (1, -1)]
        
        for row in range(self.board_size):
            for col in range(self.board_size):
                stone = self.board.get_stone(row, col)
                if stone != 0:
                    for dr, dc in directions:
                        if self._count_line(row, col, dr, dc, stone) >= 5:
                            return stone
        
        return None
    
    def _count_line(self, start_row: int, start_col: int, dr: int, dc: int, player: int) -> int:
        """Count consecutive stones in a direction."""
        count = 1  # Count the starting stone
        
        # Count forward
        r, c = start_row + dr, start_col + dc
        while (0 <= r < self.board_size and 0 <= c < self.board_size and 
               self.board.get_stone(r, c) == player):
            count += 1
            r += dr
            c += dc
        
        # Count backward
        r, c = start_row - dr, start_col - dc
        while (0 <= r < self.board_size and 0 <= c < self.board_size and 
               self.board.get_stone(r, c) == player):
            count += 1
            r -= dr
            c -= dc
        
        return count
    
    def get_board_state(self) -> List[Tuple[int, int, int]]:
        """Get current board state as list of (row, col, player) tuples."""
        return self.board.get_all_stones()
    
    def __str__(self) -> str:
        """String representation of the game state."""
        lines = [f"Move {self.move_count}, Player {self.current_player} to play"]
        lines.append(f"Captures - Player 1: {self.captures[1]}, Player -1: {self.captures[-1]}")
        lines.append(str(self.board))
        return '\n'.join(lines)

File: src/core/game_state.cpp
#include "core/game_state.hpp"
#include <algorithm>

namespace core {

MoveDelta GameState::make_move(int row, int col) noexcept {
    MoveDelta delta;
    delta.move_pos = Position{static_cast<int8_t>(row), static_cast<int8_t>(col)};
    delta.capture_count = 0;
    delta.captures_before = captures_;
    
    int player = current_player();
    
    // Place the stone
    board_.set_stone(row, col, player);
    stone_positions_.emplace_back(static_cast<int8_t>(row), static_cast<int8_t>(col));
    
    // Detect and execute captures
    detect_and_execute_captures(row, col, player, delta);
    
    // Add to move history
    move_history_.push_back(delta);
    
    // Switch player
    switch_player();
    
    return delta;
}

void GameState::undo_move() noexcept {
    if (move_history_.empty()) return;
    
    MoveDelta last_move = move_history_.back();
    move_history_.pop_back();
    
    // Switch player back
    switch_player();
    
    // Remove the placed stone
    board_.remove_stone(last_move.move_pos.row, last_move.move_pos.col);
    stone_positions_.pop_back();  // Placed stone is always the last one added
    
    // Restore captured stones
    for (uint8_t i = 0; i < last_move.capture_count; ++i) {
        Position captured = last_move.captured_stones[i];
        board_.set_stone(captured.row, captured.col, -current_player());
        stone_positions_.push_back(captured);  // Add back to position vector
    }
    
    // Restore capture counts
    captures_ = last_move.captures_before;
}

bool GameState::check_win(int row, int col, int player) const noexcept {
    // Check five in a row
    if (board_.check_five_in_a_row(row, col, player)) {
        return true;
    }
    
    // Check capture win (10 or more captures)
    return get_captures(player) >= 10;
}

bool GameState::is_terminal() const noexcept {
    if (move_history_.empty()) return false;
    
    MoveDelta last_move = move_history_.back();
    int last_player = -current_player(); // Previous player (before switch)
    
    return check_win(last_move.move_pos.row, last_move.move_pos.col, last_player);
}

int GameState::get_winner() const noexcept {
    if (!is_terminal()) return 0;
    
    MoveDelta last_move = move_history_.back();
    return -current_player(); // Previous player won
}

void GameState::detect_and_execute_captures(int row, int col, int player, MoveDelta& delta) noexcept {
    constexpr int dr[] = {0, 1, 1, 1};
    constexpr int dc[] = {1, 0, 1, -1};
    
    for (int dir = 0; dir < 4; ++dir) {
        // Check both directions from the new move
        for (int direction_sign = -1; direction_sign <= 1; direction_sign += 2) {
            // Check for capture pattern: new_move - opponent - opponent - existing_player_stone
            int r1 = row + direction_sign * dr[dir];
            int c1 = col + direction_sign * dc[dir];
            int r2 = row + direction_sign * 2 * dr[dir];
            int c2 = col + direction_sign * 2 * dc[dir];
            int r3 = row + direction_sign * 3 * dr[dir];
            int c3 = col + direction_sign * 3 * dc[dir];
            
            // Check bounds
            if (r3 < 0 || r3 >= 19 || c3 < 0 || c3 >= 19) continue;
            
            // Check capture pattern: new_move - opponent - opponent - existing_player_stone
            if (board_.get_stone(r1, c1) == -player &&
                board_.get_stone(r2, c2) == -player &&
                board_.get_stone(r3, c3) == player) {
                
                // Execute capture
                board_.remove_stone(r1, c1);
                board_.remove_stone(r2, c2);
                
                // Remove captured stones from position vector
                Position captured1{static_cast<int8_t>(r1), static_cast<int8_t>(c1)};
                Position captured2{static_cast<int8_t>(r2), static_cast<int8_t>(c2)};
                
                auto it1 = std::find(stone_positions_.begin(), stone_positions_.end(), captured1);
                if (it1 != stone_positions_.end()) {
                    stone_positions_.erase(it1);
                }
                auto it2 = std::find(stone_positions_.begin(), stone_positions_.end(), captured2);
                if (it2 != stone_positions_.end()) {
                    stone_positions_.erase(it2);
                }
                
                // Record captured positions for undo
                delta.captured_stones[delta.capture_count++] = captured1;
                delta.captured_stones[delta.capture_count++] = captured2;
                
                // Update capture count
                if (player == 1) {
                    captures_[0] += 2;
                } else {
                    captures_[1] += 2;
                }
            }
        }
    }
}

void GameState::reset() noexcept {
    board_ = BitBoard{};
    move_history_.clear();
    is_player1_turn_ = true;
    captures_ = {0, 0};
    stone_positions_.clear();
}

} // namespace core